changelog.d/12370.bugfix

```diff
@@ -0,0 +1 @@
+Fix `/messages` returning backfilled and [MSC2716](https://github.com/matrix-org/synapse/pull/12319) historic messages our of order.
```

Comment:
We can probably use the same changelog as https://github.com/matrix-org/synapse/pull/12319 as this PR doesn't actually fix anything without additionally merging that PR.

In any case, this PR does update `/messages` to do the right thing imo so it would be a good change regardless.